action/staking/stake.go

```diff
@@ -0,0 +1,163 @@
+package staking
+
+import (
+	"encoding/json"
+
+	"github.com/Oneledger/protocol/action"
+	"github.com/Oneledger/protocol/data/keys"
+	"github.com/Oneledger/protocol/identity"
+	"github.com/pkg/errors"
+	"github.com/tendermint/tendermint/libs/kv"
+)
+
+var _ action.Msg = &Stake{}
+
+type Stake struct {
+	ValidatorAddress     keys.Address
+	StakeAddress         keys.Address
+	ValidatorPubKey      keys.PublicKey
+	ValidatorECDSAPubKey keys.PublicKey
+	NodeName             string
+	Stake                action.Amount
+}
+
+func (st Stake) Marshal() ([]byte, error) {
+	return json.Marshal(st)
+}
+
+func (st *Stake) Unmarshal(data []byte) error {
+	return json.Unmarshal(data, st)
+}
+
+func (st Stake) Signers() []action.Address {
+	return []action.Address{st.StakeAddress.Bytes(), st.ValidatorAddress.Bytes()}
+}
+
+func (st Stake) Type() action.Type {
+	return action.STAKE
+}
+
+func (st Stake) Tags() kv.Pairs {
+	tags := make([]kv.Pair, 0)
+
+	tag := kv.Pair{
+		Key:   []byte("tx.type"),
+		Value: []byte(st.Type().String()),
+	}
+	tag2 := kv.Pair{
+		Key:   []byte("tx.owner"),
+		Value: st.ValidatorAddress.Bytes(),
+	}
+
+	tags = append(tags, tag, tag2)
+	return tags
+}
+
+var _ action.Tx = stakeTx{}
+
+type stakeTx struct{}
+
+func (s stakeTx) Validate(ctx *action.Context, tx action.SignedTx) (bool, error) {
+	st := &Stake{}
+	err := st.Unmarshal(tx.Data)
+	if err != nil {
+		return false, errors.Wrap(action.ErrWrongTxType, err.Error())
+	}
+	err = action.ValidateBasic(tx.RawBytes(), st.Signers(), tx.Signatures)
+	if err != nil {
+		return false, err
+	}
+
+	err = action.ValidateFee(ctx.FeePool.GetOpt(), tx.Fee)
+	if err != nil {
+		return false, err
+	}
+
+	if len(st.StakeAddress) == 0 {
```

Comment:
You can use a 'Err()' method defined in address struct which perfectly helps checking the address format.